ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java

```diff
@@ -307,10 +311,12 @@ public void awaitStop() {
 
   @Override
   public boolean addMaintainedConnectionPeer(final Peer peer) {
-    if (localNode.isReady()
-        && localNode.getPeer() != null
-        && localNode.getPeer().getEnodeURL() != null
-        && peer.getEnodeURL().getNodeId().equals(localNode.getPeer().getEnodeURL().getNodeId())) {
+    final boolean localNodeIsReady =
+        localNode.isReady()
+            && localNode.getPeer() != null
+            && localNode.getPeer().getEnodeURL() != null;
+    if (!localNodeIsReady
+        || peer.getEnodeURL().getNodeId().equals(localNode.getPeer().getEnodeURL().getNodeId())) {
```

Comment:
this change isn't needed. localNode.isReady just checks if the enode is ready (a pre-requisite to checking if our enode is the same as the peer), it's not checking if the network is ready

Reply:
^ this change is what breaks the staticNodes AcceptanceTest - so if you change this back the tests should pass :)

Reply:
Thanks @macfarla that makes sense. I've updated the check to only whether the local node is ready.

Reply:
if you revert the changes to this method addMaintainedConnectionPeer() altogether, the tests will pass and we can approve and merge it. 

localNode.isReady just means we can read the enode - needed in order to do the comparison with the peer enode

Reply:
Leaving aside the acceptance tests for a moment, the purpose of this PR is to fix a bug due to which the local node is repeatedly getting added to the maintained connection list. Here is the problem with the original code (i.e, if I revert these changes altogether):

https://github.com/hyperledger/besu/blob/main/ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java#L310

The above check _always_ fails during node startup when the caller is invoking this method for each entry present in `static-nodes.json`, which in our case happens to include the local node. As a result:

https://github.com/hyperledger/besu/blob/main/ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java#L316-L318

the _local node_ gets added added as a connection. Below are logs from a 2 node network, you can see that the peer count shows up as `2954`! I don't have the output of `admin.peers()` saved from triaging this in a live customer chain, but I did check and confirm that an entry with the local enode ID was getting duplicated hundreds of times. 


```
2022-09-19 07:18:49.072+00:00 | EthScheduler-Workers-1 | INFO  | PersistBlockTask | Imported #10,219,794 / 0 tx / 0 om / 0 (0.0%) gas / (0xe005699262c248b6b4eeb6b2b1f898a335ee4af4fb41e9b853c3d3fa07ce071b) in 0.018s. Peers: 1439
2022-09-19 07:18:52.191+00:00 | EthScheduler-Timer-0 | DEBUG | FullSyncTargetManager | Caught up to best peer: 10219793, Peers: 1439
2022-09-19 07:18:57.192+00:00 | EthScheduler-Timer-0 | DEBUG | FullSyncTargetManager | Caught up to best peer: 10219794, Peers: 1439
2022-09-19 07:18:59.045+00:00 | EthScheduler-Workers-3 | INFO  | PersistBlockTask | Imported #10,219,796 / 0 tx / 0 om / 0 (0.0%) gas / (0x05a2be65bd2b648ee081813348f09691c99d3a4e5f723eaa2f0c58acb6d4aa06) in 0.013s. Peers: 1439
2022-09-19 07:19:02.192+00:00 | EthScheduler-Timer-0 | DEBUG | FullSyncTargetManager | Caught up to best peer: 10219795, Peers: 1439
2022-09-19 07:19:04.045+00:00 | EthScheduler-Workers-0 | INFO  | PersistBlockTask | Imported #10,219,797 / 0 tx / 0 om / 0 (0.0%) gas / (0xfae06c6df283597d7ebe7e62d902544597e0cddc1aea3fb4605d3af9f954b6c4) in 0.015s. Peers: 1439
2022-09-19 07:19:07.192+00:00 | EthScheduler-Timer-0 | DEBUG | FullSyncTargetManager | Caught up to best peer: 10219796, Peers: 1439
2022-09-19 07:19:12.193+00:00 | EthScheduler-Timer-0 | DEBUG | FullSyncTargetManager | Caught up to best peer: 10219797, Peers: 1439
2022-09-19 07:19:14.060+00:00 | EthScheduler-Workers-2 | INFO  | PersistBlockTask | Imported #10,219,799 / 0 tx / 0 om / 0 (0.0%) gas / (0x606531274d1ceb438ea3abf8fcfaeb5d6a78a0f3f96d1ebbdd94fe11b940cba1) in 0.014s. Peers: 1439
...
...
...
2022-09-22 14:24:58.917+00:00 | EthScheduler-Timer-0 | DEBUG | FullSyncTargetManager | Caught up to best peer: 10276746, Peers: 2954
2022-09-22 14:24:59.734+00:00 | EthScheduler-Workers-0 | INFO  | PersistBlockTask | Imported #10,276,748 / 0 tx / 0 om / 0 (0.0%) gas / (0xcab6f9cf553508b0778475c73f5b6207680a12d97f8ad9d56ac0082ede5a7acb) in 0.489s. Peers: 2954
2022-09-22 14:25:03.952+00:00 | EthScheduler-Timer-0 | DEBUG | FullSyncTargetManager | Caught up to best peer: 10276747, Peers: 2954
2022-09-22 14:25:04.704+00:00 | EthScheduler-Workers-3 | INFO  | PersistBlockTask | Imported #10,276,749 / 0 tx / 0 om / 0 (0.0%) gas / (0x414b14b45a564bdcffe02cb9efdd4dcc98d7f9f4828a696601152814c0047078) in 0.507s. Peers: 2954
2022-09-22 14:25:08.984+00:00 | EthScheduler-Timer-0 | DEBUG | FullSyncTargetManager | Caught up to best peer: 10276748, Peers: 2954
2022-09-22 14:25:14.487+00:00 | EthScheduler-Timer-0 | DEBUG | FullSyncTargetManager | Caught up to best peer: 10276749, Peers: 2954
2022-09-22 15:22:17.029+00:00 | EthScheduler-Workers-0 | INFO  | PersistBlockTask | Imported #10,276,750 / 1 tx / 0 om / 251,662 (0.0%) gas / (0x6fdb2f23177ff7054bf27b933d113ba6dee180177dc90e9d1ade3a6cf43da23b) in 3426.331s. Peers: 2952
```

To me, this method should return `false` if any of the below holds:

* Local node is not ready - in this case, as you point out, we cannot read the enode yet.
* the entry being added is the local node.

The latest diff I've got does the above but one or more acceptance tests still fail because the tests expect to see 1 peer and there are zero peers. I'm not sure the correct solution is to revert the code.

Reply:
I think I understand what you're saying now but I think the problem is in the (regularly called) checkMaintainedConnectionPeers method rather than the addMaintained method which is only called at startup (once) or by the adminAddPeer method. I had a play around with the code and came up with this #4521 - let me know what you think

Reply:
Right, I agree that the multiple additions would occur if the list provided to `checkMaintainedConnectionPeers()` includes the local node. The change I proposed ensures that the list provided to `checkMaintainedConnectionPeers` _does not contain_ the local node whereas #4521 treats the list as potentially faulty and filters out the local node. It seems reasonable to eliminate the local node at that stage.

Given that we are going with #4521 as the fix for this issue, I will revert the changes to `addMaintainedConnectionPeer` and update the title and description.